src/puppetlabs/trapperkeeper/services_internal.clj

```diff
@@ -87,64 +94,35 @@
-             (pr-str (first (filter #(not (seq? %)) fns))))))))
-
-;; TODO Not converting to use schema since this fn's behavior should literally
-;; be a schema
-(defn validate-deps-form!
-  "Validate that the service body has a valid dependency specification.
-  Throws `IllegalArgumentException` otherwise."
-  [forms]
-  {:pre [(seq? forms)]
-   :post [(map? %)
-          (= #{:fns :dependencies} (set (keys %)))]}
-  (let [f (first forms)
-        deps (cond
-               (vector? f) f
-               (map? f) (transform-deps-map f)
-               :else (throw (IllegalArgumentException.
-                             (format
-                              "Invalid service definition; expected dependency list following protocol, found: '%s'"
-                              (pr-str f)))))]
-    (merge {:dependencies deps} (validate-fn-forms! (rest forms)))))
-
-;; TODO Not converting to use schema since this fn's behavior should literally
-;; be a schema
-(defn find-prot-and-deps-forms!
+(schema/defn ^:always-validate find-prot-and-deps-forms! :- ProtocolandDependenciesMap
   "Given the forms passed to the service macro, find the service protocol
   (if one is provided), the dependency list, and the function definitions.
   Throws `IllegalArgumentException` if the forms do not represent a valid service.
   Returns a map containing the protocol, dependency list, and fn forms."
-  [forms]
-  {:pre [(seq? forms)]
-   :post [(map? %)
-          (= #{:fns :dependencies :service-protocol-sym} (set (keys %)))
-          ((some-fn nil? symbol?) (:service-protocol-sym %))
-          (vector? (:dependencies %))
-          (seq? (:fns %))]}
-  (let [f (first forms)]
-    (cond
-      (symbol? f) (merge {:service-protocol-sym f} (validate-deps-form! (rest forms)))
-      (map? f) (merge {:service-protocol-sym nil} (validate-deps-form! forms))
-      (vector? f) (merge {:service-protocol-sym nil} (validate-deps-form! forms))
-      :else (throw (IllegalArgumentException.
+  [forms :- (schema/pred seq?)]
+  (let [f (if ((some-fn symbol? map? vector?) (first forms))
+            (first forms)
+            (throw (IllegalArgumentException.
                     (format
                      "Invalid service definition; first form must be protocol or dependency list; found '%s'"
-                     (pr-str f)))))))
+                     (pr-str (first forms))))))
+        service-protocol-sym (if (symbol? f) f nil)
+        forms (if (nil? service-protocol-sym) forms (rest forms))
+        ff (first forms)
+        deps (cond (vector? ff) ff
+                   (map? ff) (transform-deps-map ff)
+                   :else (throw (IllegalArgumentException.
+                                 (format
+                                  "Invalid service definition; expected dependency list following protocol, found: '%s'"
+                                  (pr-str ff)))))]
+    (if (every? seq? (rest forms))
```

Comment:
Given a tighter output schema, we'll now get a schema failure if (rest forms) isn't what we want. On the up side, we can avoid this check and the resulting exception; on the down side, we are trading a pretty specific error for a schema error. I feel like using schema comes with this kind of tradeoff and I'm okay with it; but @cprice404 ?